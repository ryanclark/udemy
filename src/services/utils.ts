import { InputParams, stringify } from 'query-string';

export const API_URL = 'https://gateway.marvel.com/v1/public';
export const API_KEY = '69b7e7afbbf5c89ce57b182461a97881';

export interface ApiResult<T> {
  code: number;
  status: string;
  copyright: string;
  attributionText: string;
  attributionHTML: string;
  etag: string;
  data: {
    offset: number;
    limit: number;
    total: number;
    count: number;
    results: T;
  };
}

// wrap the given URL with the Marvel hostname and the API key
export function createApiUrl(path: string, query: InputParams = {}) {
  const queryString = stringify({...query, apikey: API_KEY});

  return `${API_URL}${path}?${queryString}`;
}

// reusable method to attach the correct headers for requests and wrap the result type
export async function fetchRequest<T>(url: string): Promise<ApiResult<T>> {
  const res = await fetch(
    url,
    {
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      },
    }
  );

  return res.json();
}
