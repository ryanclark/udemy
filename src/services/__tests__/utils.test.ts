import { createApiUrl } from '../utils';

describe('Service Utils', () => {
  it('should create an API url correctly', async () => {
    const url = createApiUrl('/test', { param: 'testvalue' });

    const expected = 'https://gateway.marvel.com/v1/public/test?apikey=69b7e7afbbf5c89ce57b182461a97881&param=testvalue';

    expect(url).toBe(expected);

    expect(url).toMatchSnapshot();
  });
});
