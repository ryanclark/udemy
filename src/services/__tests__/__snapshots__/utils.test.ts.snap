// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Utils should create an API url correctly 1`] = `"https://gateway.marvel.com/v1/public/test?apikey=69b7e7afbbf5c89ce57b182461a97881&param=testvalue"`;
